[Pytorch] Specialize guts of c10::optional for 32-bit scalars

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47015 [Pytorch] Specialize guts of c10::optional for 32-bit scalars**
* #47014 [Pytorch] Add basic c10::optional tests
* #47013 [pytorch] Remove c10::nullopt_t::init

c10::optional has non-trivial copy and move operations always. This change specializes it for 32-bit scalars so that it has trivial copy and move operations in that case. Ideally, we would instead rely on P0602 "variant and optional should propagate copy/move triviality" and use `std::optional` (or implement that functionality ourselves). We can't use `std::optional` because we are stuck with C++14. Implementing the full P0602 ourselves would add even more complexity. We could do it, but this should be a helpful first step.

Differential Revision: [D24552280](https://our.internmc.facebook.com/intern/diff/D24552280/)

swolchok committed:
[Pytorch] Specialize guts of c10::optional for 32-bit scalars

c10::optional has non-trivial copy and move operations always. This change specializes it for 32-bit scalars so that it has trivial copy and move operations in that case. Ideally, we would instead rely on P0602 "variant and optional should propagate copy/move triviality" and use `std::optional` (or implement that functionality ourselves). We can't use `std::optional` because we are stuck with C++14. Implementing the full P0602 ourselves would add even more complexity. We could do it, but this should be a helpful first step.

Differential Revision: [D24552280](https://our.internmc.facebook.com/intern/diff/D24552280/)

[ghstack-poisoned]

swolchok commented:
I tried to replicate the benchmark on https://github.com/pytorch/pytorch/pull/46598 so as to provide an apples-to-apple comparison, but I couldn't run the benchmark with the most recent version of that branch checked out.

swolchok committed:
add missing bracket on "[Pytorch] Specialize guts of c10::optional for 32-bit scalars"


c10::optional has non-trivial copy and move operations always. This change specializes it for 32-bit scalars so that it has trivial copy and move operations in that case. Ideally, we would instead rely on P0602 "variant and optional should propagate copy/move triviality" and use `std::optional` (or implement that functionality ourselves). We can't use `std::optional` because we are stuck with C++14. Implementing the full P0602 ourselves would add even more complexity. We could do it, but this should be a helpful first step.

Differential Revision: [D24552280](https://our.internmc.facebook.com/intern/diff/D24552280/)

[ghstack-poisoned]

swolchok committed:
add missing #include <array> on "[Pytorch] Specialize guts of c10::optional for 32-bit scalars"


c10::optional has non-trivial copy and move operations always. This change specializes it for 32-bit scalars so that it has trivial copy and move operations in that case. Ideally, we would instead rely on P0602 "variant and optional should propagate copy/move triviality" and use `std::optional` (or implement that functionality ourselves). We can't use `std::optional` because we are stuck with C++14. Implementing the full P0602 ourselves would add even more complexity. We could do it, but this should be a helpful first step.

Differential Revision: [D24552280](https://our.internmc.facebook.com/intern/diff/D24552280/)

[ghstack-poisoned]

smessmer commented on c10/util/Optional.h:
```diff
@@ -238,23 +308,107 @@ struct constexpr_optional_base {
-    constexpr_optional_base<typename std::remove_const<
-        T>::type>, // use base with trivial destructor
-    optional_base<typename std::remove_const<T>::type>>::type;
+    std::is_scalar<T>::value && sizeof(T) <= 4,
```
`c10::ScalarType` and `c10::Layout` are enums, so they'll hit the optimization here, but we'd also like to hit the optimization for `c10::Device`, which is a 32bit struct that can for most purposes be treated like a Scalar.

Should we use checks for trivial movability/copyability here instead of `std::is_scalar`? That might hit `c10::Device` too. Also not sure if we need the 32bit sizeof constraint. If a type is trivially movable/copyable, then the optimization could make sense even if the type is larger.

We might need to rewrite the optimization a bit though. What it currently does is it unconditionally calls the copy/move constructor, possibly on uninitialized elements. With scalars, that is likely within the standard. I'm not sure if doing that for trivially copy/movable types is within the standard or UB. Do you happen to know?

swolchok replied:
> it unconditionally calls the copy/move constructor, possibly on uninitialized elements. With scalars, that is likely within the standard. I'm not sure if doing that for trivially copy/movable types is within the standard or UB. Do you happen to know?

We're not using a union, so all we require is that we can copy and move default-constructed objects of type T. If a type can't do that, it's broken. I'm not sure what's possibly uninitialized by the optional implementation; initialization is left up to T's default constructor.

swolchok replied:
> We're not using a union

Ah, and that's a problem because T may not have a default constructor. I'll see what I can do.

ezyang replied:
I feel like, for all the types we care about this optimization applying to, we have default constructors

swolchok committed:
Update on "[Pytorch] Specialize guts of c10::optional for 32-bit scalars"


c10::optional has non-trivial copy and move operations always. This change specializes it for 32-bit scalars so that it has trivial copy and move operations in that case. Ideally, we would instead rely on P0602 "variant and optional should propagate copy/move triviality" and use `std::optional` (or implement that functionality ourselves). We can't use `std::optional` because we are stuck with C++14. Implementing the full P0602 ourselves would add even more complexity. We could do it, but this should be a helpful first step.

Differential Revision: [D24552280](https://our.internmc.facebook.com/intern/diff/D24552280/)

[ghstack-poisoned]

ezyang commented:
This was a very interesting and informative PR to review, learned something. Thanks!

smessmer commented on c10/util/Optional.h:
```diff
@@ -238,23 +308,108 @@ struct constexpr_optional_base {
-        T>::type>, // use base with trivial destructor
-    optional_base<typename std::remove_const<T>::type>>::type;
+    std::is_trivially_destructible<T>::value &&
+    std::is_trivially_copyable<T>::value &&
```
I think `is_trivially_copyable` already encompasses `is_trivially_copy_assignable` and `is_trivially_move_assignable`, see https://en.cppreference.com/w/cpp/named_req/TriviallyCopyable

swolchok replied:
Right you are, fixing

smessmer commented on c10/util/Optional.h:
```diff
@@ -238,23 +308,108 @@ struct constexpr_optional_base {
 class optional : private OptionalBase<T> {
   template <class U> // re-declaration for nvcc on Windows.
   using OptionalBase = typename std::conditional<
-      std::is_trivially_destructible<U>::value, // if possible
```
We could use this PR to fix the duplication here. But not mandatory, up to you.

swolchok committed:
Update on "[Pytorch] Specialize guts of c10::optional for 32-bit scalars"


c10::optional has non-trivial copy and move operations always. This change specializes it for 32-bit scalars so that it has trivial copy and move operations in that case. Ideally, we would instead rely on P0602 "variant and optional should propagate copy/move triviality" and use `std::optional` (or implement that functionality ourselves). We can't use `std::optional` because we are stuck with C++14. Implementing the full P0602 ourselves would add even more complexity. We could do it, but this should be a helpful first step.

Differential Revision: [D24552280](https://our.internmc.facebook.com/intern/diff/D24552280/)

[ghstack-poisoned]

swolchok committed:
add extra checks for windows on "[Pytorch] Specialize guts of c10::optional for 32-bit scalars"


c10::optional has non-trivial copy and move operations always. This change specializes it for 32-bit scalars so that it has trivial copy and move operations in that case. Ideally, we would instead rely on P0602 "variant and optional should propagate copy/move triviality" and use `std::optional` (or implement that functionality ourselves). We can't use `std::optional` because we are stuck with C++14. Implementing the full P0602 ourselves would add even more complexity. We could do it, but this should be a helpful first step.

Differential Revision: [D24552280](https://our.internmc.facebook.com/intern/diff/D24552280/)

[ghstack-poisoned]

swolchok committed:
add more specific checks back in in hopes of pacifying Windows builds on "[Pytorch] Specialize guts of c10::optional for 32-bit scalars"


c10::optional has non-trivial copy and move operations always. This change specializes it for 32-bit scalars so that it has trivial copy and move operations in that case. Ideally, we would instead rely on P0602 "variant and optional should propagate copy/move triviality" and use `std::optional` (or implement that functionality ourselves). We can't use `std::optional` because we are stuck with C++14. Implementing the full P0602 ourselves would add even more complexity. We could do it, but this should be a helpful first step.

Differential Revision: [D24552280](https://our.internmc.facebook.com/intern/diff/D24552280/)

[ghstack-poisoned]

swolchok committed:
stop using std::is_trivially_copyable in Optional.cpp on "[Pytorch] Specialize guts of c10::optional for 32-bit scalars"


c10::optional has non-trivial copy and move operations always. This change specializes it for 32-bit scalars so that it has trivial copy and move operations in that case. Ideally, we would instead rely on P0602 "variant and optional should propagate copy/move triviality" and use `std::optional` (or implement that functionality ourselves). We can't use `std::optional` because we are stuck with C++14. Implementing the full P0602 ourselves would add even more complexity. We could do it, but this should be a helpful first step.

Differential Revision: [D24552280](https://our.internmc.facebook.com/intern/diff/D24552280/)

[ghstack-poisoned]
